node/src/manager/commands/copy.rs

```diff
@@ -0,0 +1,314 @@
+use diesel::{ExpressionMethods, JoinOnDsl, OptionalExtension, QueryDsl, RunQueryDsl};
+use std::{collections::HashMap, sync::Arc, time::SystemTime};
+
+use graph::{
+    components::store::BlockStore as _,
+    prelude::{
+        anyhow::{anyhow, bail, Error},
+        chrono::{DateTime, Duration, SecondsFormat, Utc},
+        ChainStore, EthereumBlockPointer, NodeId, QueryStoreManager,
+    },
+};
+use graph_store_postgres::{
+    command_support::catalog::{self, copy_state, copy_table_state},
+    PRIMARY_SHARD,
+};
+use graph_store_postgres::{connection_pool::ConnectionPool, Shard, Store, SubgraphStore};
+
+use crate::manager::deployment;
+use crate::manager::display::List;
+
+type UtcDateTime = DateTime<Utc>;
+
+#[derive(Queryable, QueryableByName, Debug)]
+#[table_name = "copy_state"]
+struct CopyState {
+    src: i32,
+    dst: i32,
+    target_block_hash: Vec<u8>,
+    target_block_number: i32,
+    started_at: UtcDateTime,
+    finished_at: Option<UtcDateTime>,
+    cancelled_at: Option<UtcDateTime>,
+}
+
+#[derive(Queryable, QueryableByName, Debug)]
+#[table_name = "copy_table_state"]
+struct CopyTableState {
+    id: i32,
+    entity_type: String,
+    dst: i32,
+    next_vid: i64,
+    target_vid: i64,
+    batch_size: i64,
+    started_at: UtcDateTime,
+    finished_at: Option<UtcDateTime>,
+    duration_ms: i64,
+}
+
+impl CopyState {
+    fn find(
+        pools: &HashMap<Shard, ConnectionPool>,
+        shard: &Shard,
+        dst: i32,
+    ) -> Result<Option<(CopyState, Vec<CopyTableState>)>, Error> {
+        use copy_state as cs;
+        use copy_table_state as cts;
+
+        let dpool = pools
+            .get(shard)
+            .ok_or_else(|| anyhow!("can not find pool for shard {}", shard))?;
+
+        let dconn = dpool.get()?;
+
+        let tables = cts::table
+            .filter(cts::dst.eq(dst))
+            .order_by(cts::entity_type)
+            .load::<CopyTableState>(&dconn)?;
+
+        Ok(cs::table
+            .filter(cs::dst.eq(dst))
+            .get_result::<CopyState>(&dconn)
+            .optional()?
+            .map(|state| (state, tables)))
+    }
+}
+
+pub async fn create(
+    store: Arc<Store>,
+    src: String,
+    src_shard: Option<String>,
+    shard: String,
+    node: String,
+    block_offset: u32,
+) -> Result<(), Error> {
+    let block_offset = block_offset as i32;
+    let subgraph_store = store.subgraph_store();
+    let src = deployment::locate(subgraph_store.as_ref(), src, src_shard)?;
+    let query_store = store.query_store(src.hash.clone().into(), true).await?;
```

Comment:
It's just that I didn't want to worry about any replication delay here when looking at the subgraph's `block_ptr`. It would probably be fine with `false`, though if there is replication lag, the error message could be pretty confusing.

Reply:
Ok